File: interviews/lc/02-best-time-stocks.cpp
#include <iostream>
#include <optional>
#include <vector>
#include "lc-utility.h"

using std::cout;
using std::optional;
using std::vector;

/**
 * maxProfit: linear algorithm with std::optional instead of scalar flag.
 * Time: O(n) Space: O(1)
 */
int maxProfit(const vector<int>& prices) {
  int profit = 0;
  optional<int> buyIdx;
  for (size_t i = 0; i < prices.size() - 1; i++) {
    if (!buyIdx.has_value() && prices[i] < prices[i + 1]) {
      buyIdx = i;
    }
    if (buyIdx.has_value() && prices[i] > prices[i + 1]) {
      profit += prices[i] - prices[*buyIdx];
      buyIdx.reset();
    }
  }
  if (buyIdx.has_value()) {
    profit += prices.back() - prices[*buyIdx];
  }
  return profit;
}

void runTestCase(vector<int> nums) {
  cout << vector2str(nums) << " = ";
  const auto res = maxProfit(nums);
  cout << res << std::endl;
}

int main(int argc, char* argv[]) {
  runTestCase({7, 1, 5, 3, 6, 4});
  runTestCase({1, 2, 3, 4, 5});
  runTestCase({7, 6, 4, 3, 1});
  return EXIT_SUCCESS;
}

File: interviews/lc/02-best-time-stocks.py
from typing import List


def maxProfit(prices: List[int]) -> int:
    buyIdx = None
    profit = 0
    for i in range(0, len(prices) - 1):
        if buyIdx is None and prices[i] < prices[i + 1]:
            buyIdx = i
        if buyIdx is not None and prices[i] > prices[i + 1]:
            profit += prices[i] - prices[buyIdx]
            buyIdx = None
    if buyIdx is not None:
        profit += prices[-1] - prices[buyIdx]
    return profit


def runTestCase(prices: List[int], expected: int) -> None:
    pricesStr = ",".join([str(p) for p in prices])
    profit = maxProfit(prices)
    print(f"{pricesStr} => Expected: {expected} Actual: {profit}")


if __name__ == "__main__":
    runTestCase([7, 1, 5, 3, 6, 4], 7)
    runTestCase([1, 2, 3, 4, 5], 4)
    runTestCase([7, 6, 4, 3, 1], 0)
